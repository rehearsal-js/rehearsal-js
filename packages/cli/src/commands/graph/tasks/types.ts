export interface GraphWorkerOptions {
  rootPath: string;
  srcPath: string;
  ignore: string[];
  serviceMap: Record<string, string>;
  output?: string;
}

type MessageResponse = { type: 'message'; content: string };
type FilesResponse = { type: 'files'; content: string[] };
export type Response = MessageResponse | FilesResponse;
