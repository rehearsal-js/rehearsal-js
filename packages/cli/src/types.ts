import { MigrationStrategy } from '@rehearsal/migration-graph';

import { UserConfig } from './userConfig';
import { State } from './helpers/state';

export type CliCommand = 'upgrade' | 'migrate';
export type Formats = 'sarif' | 'json' | 'sonarqube' | 'md';

export type MigrateCommandOptions = {
  basePath: string;
  entrypoint: string;
  files: string;
  format: Formats[];
  outputPath: string;
  verbose: boolean | undefined;
  userConfig: string | undefined;
  interactive: boolean | undefined;
};

export type MigrateCommandContext = {
  skip: boolean;
  userConfig: UserConfig | undefined;
  strategy: MigrationStrategy;
  sourceFilesWithAbsolutePath: string[];
  sourceFilesWithRelativePath: string[];
  input: unknown;
  targetPackagePath: string;
  state: State;
};

export type UpgradeCommandContext = {
  tsVersion: string;
  latestAvailableBuild: string;
  currentTSVersion: string;
  skip: boolean;
};

export type UpgradeCommandOptions = {
  build: string;
  basePath: string;
  format: Formats[];
  outputPath: string;
  tsVersion: string;
  dryRun: boolean;
};

export type DependencyConfig = {
  dependencies?: string[];
  devDependencies?: string[];
};

export type SetupConfigCommand = {
  command: string;
  args: string[];
};

export type SetupConfig = {
  ts?: SetupConfigCommand;
  lint?: SetupConfigCommand;
};

export type CustomCommandConfig = {
  install?: DependencyConfig;
  setup?: SetupConfig;
};

export type CustomConfig = {
  upgrade?: CustomCommandConfig;
  migrate?: CustomCommandConfig;
};

export type PackageSelection = {
  name: string;
  path: string;
};

export type MenuMap = {
  [key: string]: string;
};

export type MigrationSummary = {
  totalErrorCount: number;
  hintAddedCount: number;
};
export type TSConfig = {
  compilerOptions: {
    strict: boolean;
  };
};

export type ScriptMap = {
  [key: string]: string;
};
