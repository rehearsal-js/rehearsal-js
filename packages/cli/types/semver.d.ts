declare module 'semver' {
  const semver: {
    valid: (input: string) => boolean;
  };
  export = semver;
}
