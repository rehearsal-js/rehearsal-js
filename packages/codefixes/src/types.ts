import {
  CodeFixAction,
  DiagnosticWithLocation,
  FileTextChanges,
  LanguageService,
  Node,
  Program,
  TypeChecker,
} from 'typescript';

export interface DiagnosticWithContext extends DiagnosticWithLocation {
  service: LanguageService;
  program: Program;
  checker: TypeChecker;
  node?: Node;
}

export interface CodeHintList {
  [key: number]: CodeHintListElement;
}

export interface CodeHintListElement {
  hint: string;
  helpUrl?: string;
  hints?: CodeHint[];
}

export interface CodeHint {
  when: (n: Node, p: Program, c: TypeChecker) => boolean;
  hint: string;
}

export interface CodeFixCollection {
  getFixForDiagnostic(diagnostic: DiagnosticWithContext): CodeFixAction | undefined;
}

export type CodeFixKind = 'insert' | 'delete' | 'replace';

export interface FixedFile {
  fileName: string;
  newCode?: string;
  oldCode?: string;
  codeFixAction?: CodeFixKind;
  location: {
    startLine: number;
    startColumn: number;
    endLine: number;
    endColumn: number;
  };
}

export interface CodeFix {
  getCodeAction: (diagnostic: DiagnosticWithContext) => CodeFixAction | undefined;
}

export function createCodeFixAction(
  fixName: string,
  changes: FileTextChanges[],
  description: string
): CodeFixAction {
  return {
    fixName,
    description,
    changes,
  };
}
