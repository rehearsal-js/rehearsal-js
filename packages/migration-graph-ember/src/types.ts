/* eslint-disable @typescript-eslint/no-explicit-any */
import type { Package } from '@rehearsal/migration-graph-shared';
import type { EmberAddonPackage } from './entities/ember-addon-package.js';
import type { EmberAppPackage } from './entities/ember-app-package.js';

export type EmberProjectPackage = Package | EmberAddonPackage | EmberAppPackage;
