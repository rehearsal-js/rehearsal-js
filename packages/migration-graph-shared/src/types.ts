import type { Package } from './entities/package.js';

export type UniqueNode = {
  key: string;
  synthetic?: boolean;
};

export type PackageNode = UniqueNode & {
  pkg: Package;
  converted?: boolean;
};

export type ModuleNode = UniqueNode & {
  path: string;
  parsed?: boolean;
};
