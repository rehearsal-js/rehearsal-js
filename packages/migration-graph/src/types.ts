import { Package } from '@rehearsal/migration-graph-shared';

import { Graph } from './utils/graph';

export type UniqueGraphNode = {
  key: string;
};

export type PackageNode = UniqueGraphNode & {
  pkg: Package;
  converted?: boolean;
  modules: Graph<ModuleNode>;
};

export type ModuleNode = UniqueGraphNode & {
  path: string;
  parsed?: boolean;
  meta?: unknown;
};
