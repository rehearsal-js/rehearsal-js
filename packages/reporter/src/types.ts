export type Formatters = 'json' | 'sonarqube' | 'md' | 'sarif';

export type ReportSummary = Record<string, unknown> & {
  projectName: string;
  tsVersion: string;
  timestamp: string;
};

export interface Location {
  startLine: number;
  startColumn: number;
  endLine: number;
  endColumn: number;
}

export enum ReportItemType {
  ts = 0,
  lint = 1,
  glint = 2,
}

export type ReportItem = {
  analysisTarget: string;
  type: ReportItemType;
  ruleId: string;
  category: string;
  message: string;
  hint?: string;
  hintAdded?: boolean;
  nodeKind?: string;
  nodeText?: string;
  helpUrl?: string;
  nodeLocation?: {
    startLine: number;
    startColumn: number;
    endLine: number;
    endColumn: number;
  };
};

export interface LintErrorLike {
  message: string;
  ruleId: string | null;
  line?: number;
  column?: number;
  nodeType?: string;
  messageId?: string;
  endLine?: number;
  endColumn?: number;
}

export type Report = {
  summary: ReportSummary[];
  fixedItemCount: number;
  items: ReportItem[];
};

export type ReportFormatter = (report: Report) => string;

export interface Run {
  runSummary: ReportSummary;
  fixedItemCount: number;
  items: ReportItem[];
}

// ts doesnt allow for static properties on interfaces yet
export class FormatterBase {
  static extension: string;
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  static getReport(_report: Report): string {
    return '';
  }
}
