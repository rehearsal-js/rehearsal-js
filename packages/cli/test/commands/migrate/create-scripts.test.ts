import { resolve } from 'node:path';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { writeJSONSync } from 'fs-extra/esm';

import { createScriptsTask } from '../../../src/commands/migrate/tasks/index.js';
import { prepareTmpDir, listrTaskRunner, createMigrateOptions } from '../../test-helpers/index.js';
import { PackageJson } from '../../../src/types.js';

describe('Task: create-scripts', () => {
  let basePath = '';
  let output = '';
  vi.spyOn(console, 'info').mockImplementation((chunk) => {
    output += `${chunk}\n`;
  });
  vi.spyOn(console, 'log').mockImplementation((chunk) => {
    output += `${chunk}\n`;
  });

  beforeEach(() => {
    output = '';
    basePath = prepareTmpDir('basic');
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test('add lint:tsc in package.json', async () => {
    const options = createMigrateOptions(basePath);
    const tasks = [createScriptsTask(options)];
    await listrTaskRunner(tasks);

    const packageJson = PackageJson.parse(resolve(basePath, 'package.json'));
    expect(packageJson?.scripts?.['lint:tsc']).toBe('tsc --noEmit');

    expect(output).matchSnapshot();
  });

  test('replace if the script exists', async () => {
    const options = createMigrateOptions(basePath);
    const tasks = [createScriptsTask(options)];
    await listrTaskRunner(tasks);

    const packageJson = PackageJson.parse(resolve(basePath, 'package.json'));
    writeJSONSync(resolve(basePath, 'package.json'), {
      scripts: { 'lint:tsc': 'foo' },
      ...packageJson,
    });
    expect(packageJson?.scripts?.['lint:tsc']).toBe('tsc --noEmit');

    expect(output).matchSnapshot();
  });
});
