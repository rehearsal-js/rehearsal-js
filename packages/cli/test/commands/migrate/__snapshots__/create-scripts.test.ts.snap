// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Task: create-scripts > add lint:tsc in package.json 1`] = `
"[STARTED] Add package scripts
[DATA] Adding \\"lint:tsc\\" script in package.json
[SUCCESS] Add package scripts
"
`;

exports[`Task: create-scripts > replace if the script exists 1`] = `
"[STARTED] Add package scripts
[DATA] Adding \\"lint:tsc\\" script in package.json
[SUCCESS] Add package scripts
"
`;
